fixed #1 Refactor SelectionSort

JSiggelkow opened the pull request:
Umstellung von BubbleSort-artigem Verhalten auf korrektes SelectionSort Verfahren

Die SelectionSort-Erklärung im Skript hat ebenfalls das BubbleSort-artige Verhalten...

Commit:
Refactor SelectionSort

Umstellung von BubbleSort-artigem Verhalten auf korrektes SelectionSort Verfahren

NErzberger commented:
Hallo. Es macht tatsächlich keinen Unterschied. Deine implementierten ist genauso richtig. Es erweckt den Anschein, dass sich der Algorithmus ähnlich wie der bubble sort verhält, da in meiner Implementierung alles einmal durchgetauscht wird, bis das kleinste Element gefunden wird. Aber Schluss endlich wird von vorne her das Array aufsteigend sortiert. In deinem Vorschlag werden lediglich weniger Tausch Aktionen durchgeführt. Beide Algorithmen verhalten sich aber gleich.

JSiggelkow commented:
Ja, natürlich sortieren beide Algorithmen das Array. Jedoch geht es doch in der Aufgabe darum das Prinzip des SelectionSorts zu verstehen und in der vorherigen Implementierung wird das Prinzip des SelectionSorts nicht verfolgt. 
In deiner Implementierung tauschst du alle einmal durch, wie du selbst gesagt hast... aber das ist ja dann kein SelectionSort mehr:

Prinzip des Selectionsort:
Sei S der sortierte Teil des [Arrays](vorne im Array) und U der unsortierte Teil (dahinter). Am Anfang ist S noch leer, U entspricht dem ganzen (restlichen) Array. Das Sortieren durch Auswählen läuft nun folgendermaßen ab:

Suche das kleinste Element in U und vertausche es mit dem ersten Element von U (= das erste Element nach S).

Danach ist das Array bis zu dieser Position sortiert. Das kleinste Element wird in S verschoben (indem S einfach als ein Element länger betrachtet wird, und U nun ein Element später beginnt). S ist um ein Element gewachsen, U um ein Element kürzer geworden. Anschließend wird das Verfahren so lange wiederholt, bis das gesamte Array abgearbeitet worden ist; S umfasst am Ende das gesamte Array, aufsteigend sortiert, U ist leer.

von: https://de.wikipedia.org/wiki/Selectionsort

NErzberger commented:
Deine Erklärung ist vollkommen richtig. Jedoch hast du eine Sache nicht ganz verstanden. Nämlich meine Implementierung.

Meine Implementierung geht her und tauscht das nächste kleiner Element mit dem an der Stelle I. Aber I bleibt stehen. Dann steht an der Stelle I ein kleineres Element als an Stelle j und j wird weiter hochgezählt, bis wieder ein kleiners Element kommt als an Stelle I.
Was ich gemeint habe: es werden alle kleinere Elemente mit dem am stelle I getauscht, bis an Stelle I das kleinste Element ist.

Es soll die 7 sortiert werden

2 7 5 8 3

Erster Schritt der inneren Schleife:
2 5 7 8 3


Dritter Schritt der inneren Schleife, da der zweite nix gemacht hat:
2 3 7 8 5


Dort, wo du der einfach den Index vom kleinsten Element merkst, tausche ich die Elemente.
Ansich macht das überhaupt keinen Unterschied.

Daher sind beide Algorithmen bzw. Implementierten richtig.

JSiggelkow commented:
Ja, das ist ja richtig was du sagst und ich habe verstanden wie deine Implementierung funktioniert.
Jedoch ist es wie gesagt nicht das Verhalten, was bei einem SelectionSort erwartet wird.
Bei dem SelectionSort soll immer das kleinste Element der verbleibenden Liste gefunden werden und **nichts** anderes.

Zu deinem Beispiel:
Wir wollen die 7 sortieren:

2 7 5 8 3 

Dein Algorithmus würde dann die 7 mit der 5 Tauschen und dann im nächsten durchlauf die 5 mit der 3
Dein Array sieht dann wie folgt aus: 2 3 7 8 5

Die Idee des SelectionSort ist es jedoch das kleinste Element zu finden und dann halt nur die 3 mit der 7 zu Tauschen
Ein Array mit dem SelectionSort sortiert sieht dann so aus: 2 3 5 8 7

Wie man sieht gibt es halt schon einen Unterschied.
Natürlich sortieren die beiden Algorithmen das Array aber was ich halt mit dem Issue sagen möchte, ist das z.B. das Tauschen der 7 mit der 5, obwohl es noch eine 3 im verbleibenden unsortierten Bereich des Arrays gibt einfach nicht dem Prinzip des SelectionSort entspricht und meiner Meinung für die Aufgabe "Implementieren Sie den SelectionSort" falsch ist.

NErzberger commented:
Das Beispiel ist tatsächlich etwas blöd. Der Algorithmus entspricht jedoch trotzdem dem Grundsatz des SelectionSort, dass am Ende das kleinste Element vorne steht.

Ich nehme den Punkt mit und werde es abändern.

Die Implementierung ist in der Klausur trotzdem richtig.

Commit:
Fix #4: Division durch Null in der Interpolationssuche verhindern

Dieser Commit behebt ein Division-durch-Null-Problem in der Implementierung der Interpolationssuche. Es wurde eine Prüfung hinzugefügt, die überprüft, ob der Wert an der linken Grenze dem Wert an der rechten Grenze entspricht. Da das Array sortiert ist, können wir, wenn diese Werte gleich sind, davon ausgehen, dass auch alle Werte dazwischen gleich sind.

Nach dieser Prüfung gibt die Funktion entweder die Position der linken Grenze zurück (wenn der Wert an der linken Grenze dem gesuchten Wert entspricht), oder sie gibt -1 zurück, um anzuzeigen, dass der gesuchte Wert im Array nicht gefunden werden konnte.
